### ISSUE TITLE
Can someone please answer some VERY basic questions I have. :(

### ISSUE THREAD

**Issue by pGit1** [tloen/alpaca-lora#162]
So to any interested parties, especially @tloen , I have some rather *dumb* questions I'd rather NOT handwave away. 

1. What exactly is going on in finetune.py. I am not using the latest version, I am using this version:

```
def generate_prompt(data_point):
    # sorry about the formatting disaster gotta move fast
    if data_point["input"]:
        return f"""Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{data_point["instruction"]}

### Input:
{data_point["input"]}

### Response:
{data_point["output"]}"""
    else:
        return f"""Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
{data_point["instruction"]}

### Response:
{data_point["output"]}"""


def tokenize(prompt):
    # there's probably a way to do this with the tokenizer settings
    # but again, gotta move fast
    result = tokenizer(
        prompt,
        truncation=True,
        max_length=CUTOFF_LEN + 1,
        padding="max_length",
    )
    return {
        "input_ids": result["input_ids"][:-1],
        "attention_mask": result["attention_mask"][:-1],
    }


def generate_and_tokenize_prompt(data_point):
    prompt = generate_prompt(data_point)
    return tokenize(prompt)


if VAL_SET_SIZE > 0:
    train_val = data["train"].train_test_split(
        test_size=VAL_SET_SIZE, shuffle=True, seed=42
    )
    train_data = train_val["train"].shuffle().map(generate_and_tokenize_prompt)
    val_data = train_val["test"].shuffle().map(generate_and_tokenize_prompt)
else:
    train_data = data["train"].shuffle().map(generate_and_tokenize_prompt)
    val_data = None
```

Not being a huggingface user prior to this repo I have been interrogating variables to see what is going on, and it makes no sense to me. :( 

For instance after letting the code run to the point of data creation I see the output below when looking at `val_data` or `train_data`: 
```
Dataset({
    features: ['instruction', 'input', 'output', 'input_ids', 'attention_mask'],
    num_rows: 2000
})
```

BUT how on earth does this `transformers.Trainer` class know what to do with this dictionary object? At first I thought by "finetuning" we were doing more "next word prediction" and I started mastering the math of why it works during training and how the various sampling algorithms work during inference. I started feeling good and then realized I actually HAVE NO IDEA, how this "supervised" finetuning works. Given that instructions and outputs in the json data are different, I am guessing we arent doing finetuning with next word prediction? Or are we concatenating "instruction" and "input" together and we treat "output" as labels but still doing the causal llm task of next word prediction? So confused here. I will say this repo is **awesome** and that Ive successfully finetuned a model on a **single gpu** which is insane to me, but refuse to use the model since I dont really know what is going on in the finetuning step. :(

Really amazed how the `transformers.Trainer` class just handles this arbitrary key/value store without blinking and I cant find anything in the Trainer source code that allows it to do this. :( Overall I just have no idea what is going on and I thought I did before today.

**Comment by tloen** [comment]
What's important is that the dataset has the keys `input_ids` and `attention_mask`. These are the keys used by the the `Trainer` class, as you can see e.g. here:

https://github.com/huggingface/transformers/blob/68287689f2f0d8b7063c400230b3766987abf18d/src/transformers/trainer.py#L3114

**Comment by pGit1** [comment]
@tloen Thanks for answering and that makes alot more sense. Can you comment on what the actual finetuning step is doing? This part is not clear to me. Are we doing the next step prediction or something else?

**Comment by tloen** [comment]
Given a token sequence, the Trainer is just computing the modeled probability of every token in the sequence and then using gradient descent to tweak the model to increase those probabilities. Repeat for every sequence in the dataset, three times, and you get the finetuned model.

**Comment by pGit1** [comment]
So I think I got. I much prefer math or diagrams but I think this makes almost trivial sense now. 

Given some 

```Batch, seq_len, dim`` input tensor the model spits out a ```Batch, seq_len, vocab``` logits prediction tensor and the softmax of the logits is compared to the ground truth which is ```Batch, seq_len, vocab``` one hots, corresponding to the desired outputs. If this isnt correct than I completely lost. Im assuming this is correct.

Last question then, are the "instruction" AND "input" used to train the model as inputs?

Thanks so much for answering!

**Comment by tloen** [comment]
That's more or less correct. Technically `dim` is 1 here, unless you've already gathered the embeddings.

> Last question then, are the "instruction" AND "input" used to train the model as inputs?

They're an artifact from the dataset mapping, not used in the Trainer.

**Comment by pGit1** [comment]
Yes indeed. ```dim``` was meant to be general to indicate whether or not we were ingesting the integers from tokenizer or token embeddings. 

@tloen I am SO SORRY for asking these silly questions but if 

```
{
        "instruction": "Identify the odd one out.",
        "input": "Twitter, Instagram, Telegram",
        "output": "Telegram"
    },
``` 
"input" and "instruction" are not used by Trainer class what on earth is it using for inputs to the model?

**Comment by tloen** [comment]
They're being tokenized into the input_ids.

**Comment by pGit1** [comment]
Perfect! Gotcha. I can use the finetuned model now :sweat_smile: 

THANKS!

**Comment by pGit1** [comment]
@tloen I think my understanding above and your original code that I posted above might be incompatible on further inspection, unless I am mistake.

When I inspect these variables after running the code:
```
train_dataloader = trainer.get_train_dataloader()
# get an example minibatch
example_minibatch = next(iter(train_dataloader))
# This looks exaclty like
example_minibatch['labels']
# This :(
example_minibatch['input_ids']
```

the labels and input_ids in the minibatch look identical. I expected given the collate function it would "shift labels to the right by one" per the hugginface docs but this isnt the case. I am not sure what is going on here. Shouldn't labels be the "output" field from the json files? Second it looks like the minibatches are 1to1 so how is this handled in trainer? Are we actually doing next work prediction?

**Comment by pGit1** [comment]
Now that I get hugginface :angry: I am going after the 13b parameter model now... :)

**Comment by disperaller** [comment]
> Yes indeed. `dim` was meant to be general to indicate whether or not we were ingesting the integers from tokenizer or token embeddings.
> 
> @tloen I am SO SORRY for asking these silly questions but if
> 
> ```
> {
>         "instruction": "Identify the odd one out.",
>         "input": "Twitter, Instagram, Telegram",
>         "output": "Telegram"
>     },
> ```
> 
> "input" and "instruction" are not used by Trainer class what on earth is it using for inputs to the model?

why is the dim being 1 here, in the traditional Transformer architecture, shouldn't it being the embedding size for each token?

**Comment by pGit1** [comment]
Look at the generate_prompt code. What it does is that it takes
instruction, input and output and converts it to one long string, that then
gets tokenized and becomes input_ids. Trainer class uses input_ids.
